temporalio/workflow.py

```diff
@@ -232,13 +279,12 @@ def with_name(
             )
         return fn
 
-    if name is not None or dynamic:
+    if not fn:
         if name is not None and dynamic:
             raise RuntimeError("Cannot provide name and dynamic boolean")
-        return partial(with_name, name)
-    if fn is None:
-        raise RuntimeError("Cannot create signal without function or name or dynamic")
-    return with_name(fn.__name__, fn)
+        return partial(decorator, name, unfinished_handlers_policy)
+    else:
+        return decorator(fn.__name__, unfinished_handlers_policy, fn)
```

Comment:
These changes make the decorator behave consistently: permitting e.g. `@update()` with the same meaning as `@update`, and allowing the default value of the handler policy to be passed explicitly as a param if the user desires. That makes sense, especially now that we're adding a 3rd parameter to the decorators which is unrelated to `name` and `dynamic`. If we go ahead with this, I'll need to update the query decorator also for consistency.